utxo.go

```diff
@@ -0,0 +1,198 @@
+		}
+
+		output, err := txhelper.MakeTxOutput(destination.Address, destination.AtomAmount, tx.sourceWallet.chainParams)
+		if err != nil {
+			return nil, 0, "", fmt.Errorf("make tx output error: %v", err)
+		}
+
+		totalSendAmount += output.Value
+		outputs = append(outputs, output)
+	}
+
+	return outputs, totalSendAmount, maxAmountRecipientAddress, nil
+}
+
+func (tx *TxAuthor) constructCustomTransaction() (*txauthor.AuthoredTx, error) {
+	// Used to generate an internal address for change,
+	// if no change destination is provided and
+	// no recipient is set to receive max amount.
+	nextInternalAddress := func() (string, error) {
+		ctx := tx.sourceWallet.shutdownContext()
+		addr, err := tx.sourceWallet.internal.NewChangeAddress(ctx, tx.sourceAccountNumber)
+		if err != nil {
+			return "", err
+		}
+		return addr.Address(), nil
+	}
+	var totalInputAmount int64
+
+	msgTx, maxSignedSize, err := func(inputs []*wire.TxIn, sendDestinations, changeDestinations []TransactionDestination,
+		nextInternalAddress NextAddressFunc) (*wire.MsgTx, int, error) {
+		outputs, totalSendAmount, maxAmountRecipientAddress, err := tx.ParseOutputsAndChangeDestination(sendDestinations)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		if maxAmountRecipientAddress != "" && len(changeDestinations) > 0 {
+			return nil, 0, errors.E(errors.Invalid, "no change is generated when sending max amount,"+
+				" change destinations must not be provided")
+		}
+
+		if maxAmountRecipientAddress == "" && len(changeDestinations) == 0 {
+			// no change specified, generate new internal address to use as change (max amount recipient)
+			maxAmountRecipientAddress, err = nextInternalAddress()
+			if err != nil {
+				return nil, 0, fmt.Errorf("error generating internal address to use as change: %s", err.Error())
+			}
+		}
+
+		inputScriptSizes := make([]int, len(inputs))
+		inputScripts := make([][]byte, len(inputs))
+		for i, input := range inputs {
+			totalInputAmount += input.ValueIn
+			inputScriptSizes[i] = txsizes.RedeemP2PKHSigScriptSize
+			inputScripts[i] = input.SignatureScript
+		}
+
+		var changeScriptSize int
+		if maxAmountRecipientAddress != "" {
+			changeScriptSize, err = calculateChangeScriptSize(maxAmountRecipientAddress, tx.sourceWallet.chainParams)
+		} else {
+			changeScriptSize, err = calculateMultipleChangeScriptSize(changeDestinations, tx.sourceWallet.chainParams)
+		}
+		if err != nil {
+			return nil, 0, err
+		}
+
+		maxSignedSize := txsizes.EstimateSerializeSize(inputScriptSizes, outputs, changeScriptSize)
+		maxRequiredFee := txrules.FeeForSerializeSize(txrules.DefaultRelayFeePerKb, maxSignedSize)
+		changeAmount := totalInputAmount - totalSendAmount - int64(maxRequiredFee)
+
+		if changeAmount < 0 {
+			excessSpending := 0 - changeAmount // equivalent to math.Abs()
+			return nil, 0, fmt.Errorf("total send amount plus tx fee is higher than the total input amount by %s",
+				dcrutil.Amount(excessSpending).String())
```

Comment:
```suggestion
			return nil, 0, errors.New(ErrInsufficientBalance)
```

Since you're sending using custom inputs, I assume you'll handle this error in the frontend as insufficient inputs, not insufficient balance.

Reply:
Should I add more new of errors type, like `ErrInsufficientInputsBalance`?